File: boot-redis/src/main/java/com/example/demo/SecurityConfig.java
package com.example.demo;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.web.server.ServerHttpSecurity;
import org.springframework.security.web.server.SecurityWebFilterChain;
import org.springframework.security.web.server.context.NoOpServerSecurityContextRepository;

@Configuration
public class SecurityConfig {

    @Bean
    public SecurityWebFilterChain springWebFilterChain(ServerHttpSecurity http) {
        return http.csrf(ServerHttpSecurity.CsrfSpec::disable)
                .httpBasic(it -> it.securityContextRepository(NoOpServerSecurityContextRepository.getInstance()))
                .authorizeExchange(it -> it.pathMatchers("/posts/**").authenticated()
                        .anyExchange().permitAll())
//                        .pathMatchers(HttpMethod.POST, "/posts/**").authenticated()
//                        .pathMatchers(HttpMethod.DELETE, "/posts/**").authenticated()
//                        .anyExchange().permitAll())
                .build();
    }
}

File: kotlin-gradle/src/main/kotlin/com/example/demo/SecurityConfig.kt
package com.example.demo

import org.springframework.context.annotation.Bean
import org.springframework.http.HttpMethod
import org.springframework.security.authorization.AuthorizationDecision
import org.springframework.security.config.annotation.web.reactive.EnableWebFluxSecurity
import org.springframework.security.config.web.server.HttpSecurity
import org.springframework.security.core.Authentication
import org.springframework.security.core.userdetails.MapUserDetailsRepository
import org.springframework.security.core.userdetails.User
import org.springframework.security.web.server.SecurityWebFilterChain
import org.springframework.security.web.server.authorization.AuthorizationContext
import reactor.core.publisher.Mono

@EnableWebFluxSecurity
class SecurityConfig {
    @Bean
    @Throws(Exception::class)
    fun springWebFilterChain(http: HttpSecurity): SecurityWebFilterChain {
        return http
                .authorizeExchange()
                .pathMatchers(HttpMethod.GET, "/posts/**").permitAll()
                .pathMatchers(HttpMethod.DELETE, "/posts/**").hasRole("ADMIN")
                //.pathMatchers("/users/{user}/**").access(this::currentUserMatchesPath)
                .anyExchange().authenticated()
                .and()
                .build()
    }

    private fun currentUserMatchesPath(authentication: Mono<Authentication>, context: AuthorizationContext): Mono<AuthorizationDecision> {
        return authentication
                .map { context.variables?.get("user")?.equals(it.name) }
                .map { AuthorizationDecision(it ?: false) }
    }

    @Bean
    fun userDetailsRepository(): MapUserDetailsRepository {
        val rob = User.withUsername("test").password("test123").roles("USER").build()
        val admin = User.withUsername("admin").password("admin123").roles("USER", "ADMIN").build()
        return MapUserDetailsRepository(rob, admin)
    }
}